### PR TITLE
DRAFT: POC omni pool manager

### PR DESCRIPTION
**Pull request by 0xahtle7**

### CODE REVIEW

File: contracts/sidechain/L1PoolManagerProxy.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * @notice Send a message to add a pool on a sidechain.
+     * @param _gauge              The root gauge address.
+     * @param _dstChainId         The LayerZero destination chain ID eg optimism is 111
+     * @param _zroPaymentAddress  The LayerZero ZRO payment address
+     * @param _adapterParams      The adapter params
+     */
+    function addPool(
+        address _gauge,
+        uint16 _dstChainId,
```
Comment (phijfry) [inline contracts/sidechain/L1PoolManagerProxy.sol:88]: How can you validate this is the correct chain ID?

File: contracts/sidechain/L1PoolManagerProxy.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.11;
+
+import { LzApp } from "../layerzero/lzApp/LzApp.sol";
+
+interface IStakelessGauge {
+    function getRecipient() external view returns (address);
+}
+
+interface IGaugeController {
```
Comment (phijfry) [inline contracts/sidechain/L1PoolManagerProxy.sol:10]: Try and keep interfaces in a separate file in Interfaces folder where possible

File: contracts/sidechain/L1PoolManagerProxy.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        address dstGauge = IStakelessGauge(_gauge).getRecipient();
+        require(dstGauge != address(0), "!dstGauge");
+        //check that the pool as weight
+        uint256 weight = IGaugeController(gaugeController).get_gauge_weight(_gauge);
+        require(weight > 0, "must have weight");
+
+        _lzSend(
+            _dstChainId, ///////////// Destination chain (L2 chain)
+            abi.encode(dstGauge), ///////////////// Payload encode
+            payable(address(this)), // Refund address
```
Comment (phijfry) [inline contracts/sidechain/L1PoolManagerProxy.sol:106]: Refund address should probably be the sender, as this is possibly publicly callable the sender will loose their refund.

File: contracts/sidechain/L2PoolManagerProxy.sol [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    function setPoolManagerOperator(address _operator) external onlyOwner {
+        require(address(0) != _operator, "!_operator");
+        IPoolManager(poolManager).setOperator(_operator);
+    }
+
+    /**
+     * @notice Adds new pool.
+     * @param _gauge The gauge address.
+     */
+    function addPool(address _gauge) external returns (bool) {
```
Comment (phijfry) [inline contracts/sidechain/L2PoolManagerProxy.sol:71]: Isn't this just making it possible for anybody to add any gauge?

File: test/sidechain/L1PoolManagerProxy.spec.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            const tx = await l1PoolManagerProxy
+                .connect(dao.signer)
+                .addPool(rootGauge.address, L2_CHAIN_ID, ZERO_ADDRESS, adapterParams, { value: NATIVE_FEE });
+            await expect(tx)
+                .to.emit(l1PoolManagerProxy, "AddSidechainPool")
+                .withArgs(L2_CHAIN_ID, rootGauge.address, sidechainGauge.address);
+            await expect(tx).to.emit(l1PoolManagerProxy, "PoolAdded");
+        });
+    });
+    // describe("edge cases", () => {
```
Comment (phijfry) [inline test/sidechain/L1PoolManagerProxy.spec.ts:173]: Commented code 😱😂

File: contracts/sidechain/L1PoolManagerProxy.sol (hunk 3 above)
Comment (0xahtle7) [reply contracts/sidechain/L1PoolManagerProxy.sol]: thanks , i changed it

File: contracts/sidechain/L1PoolManagerProxy.sol (hunk 1 above)
Comment (0xahtle7) [reply contracts/sidechain/L1PoolManagerProxy.sol:93]: via the gaugeCheckpointer

**Review by phijfry — APPROVED** [review]
Looks good to me, Balancer also gave it the h/t of approval in terms of our use of the gauge checkpointer.

File: contracts/sidechain/L2PoolManagerProxy.sol [hunk 6]
```diff
@@ -44,12 +67,17 @@ contract L2PoolManagerProxy is NonblockingLzApp {
...
+       Core functions
+    ------------------------------------------------------------------- */
+
     /**
      * @notice Adds new pool directly on L2.
      * @param _gauge The gauge address.
      */
-    function addPool(address _gauge) external onlyOwner returns (bool) {
-        return _addPool(_gauge);
+    function addPool(address _gauge) external onlyKeeper returns (bool) {
```
Comment (0xahtle7) [inline contracts/sidechain/L2PoolManagerProxy.sol:78]: We still a version of this pools that is onlyOwner and does not need to check `isValidGauge`  to for example being able to add this https://app.aura.finance/#/8453/pool/11 or similar

File: contracts/sidechain/L2PoolManagerProxy.sol [hunk 7]
```diff
@@ -3,24 +3,35 @@ pragma solidity 0.8.11;
...
 import { IPoolManagerLite } from "contracts/sidechain/interfaces/IPoolManagerLite.sol";
+import { KeeperRole } from "../peripheral/KeeperRole.sol";
 
 /**
  * @title   L2PoolManagerProxy
  * @author  AuraFinance
  * @dev     Given a root gauge on L1PoolManagerProxy it adds a gauge recipient on PoolManagerLite
  */
-contract L2PoolManagerProxy is NonblockingLzApp {
+contract L2PoolManagerProxy is NonblockingLzApp, KeeperRole {
```
Comment (0xahtle7) [inline contracts/sidechain/L2PoolManagerProxy.sol:13]: The KeeperRole  passes on the constructor the msg.sender what will make this contract to have different address on each chain.

Comment (phijfry) [reply contracts/sidechain/L2PoolManagerProxy.sol:13]: should deploy with the same address no? Assuming msg.sender is the same which it will be because its the factory:

`address = keccak256(0xff ++ sender_address ++ salt ++ keccak256(init_code))[12:]`

**Review by phijfry — APPROVED** [review]
